File: wrappers/vb/LibRangeMap/TemporalRangeMapper.vb
Imports System.Text.Json
Imports System.Text.Json.Serialization

Public NotInheritable Class TemporalRangeMapper
        Private Shared ReadOnly JsonOptions As New JsonSerializerOptions With {
            .PropertyNamingPolicy = Nothing,
            .WriteIndented = False
        }

        Private Const SpecType As String = "temporal_range"

        Public ReadOnly Property InputMin As Long
        Public ReadOnly Property InputMax As Long
        Public ReadOnly Property OutputMin As Double
        Public ReadOnly Property OutputMax As Double
        Public ReadOnly Property Clip As Boolean

        Public Sub New(
            inputMinUnixMilliseconds As Long,
            inputMaxUnixMilliseconds As Long,
            Optional outputMin As Double = -1.0,
            Optional outputMax As Double = 1.0,
            Optional clip As Boolean = False
        )
            If inputMaxUnixMilliseconds <= inputMinUnixMilliseconds Then
                Throw New ArgumentException("input_min must be less than input_max.")
            End If
            If outputMax <= outputMin Then
                Throw New ArgumentException("output_min must be less than output_max.")
            End If

            Me.InputMin = inputMinUnixMilliseconds
            Me.InputMax = inputMaxUnixMilliseconds
            Me.OutputMin = outputMin
            Me.OutputMax = outputMax
            Me.Clip = clip
        End Sub

        Public Function MapValue(value As Long) As Double
            Return MapUnixMilliseconds(value)
        End Function

        Public Function MapValue(value As DateTime) As Double
            If value.Kind = DateTimeKind.Unspecified Then
                Throw New ArgumentException("DateTimeKind.Unspecified is not supported; use UTC, Local, or DateTimeOffset.")
            End If

            Return MapUnixMilliseconds(New DateTimeOffset(value).ToUnixTimeMilliseconds())
        End Function

        Public Function MapValue(value As DateTimeOffset) As Double
            Return MapUnixMilliseconds(value.ToUnixTimeMilliseconds())
        End Function

        Public Function Spec() As TemporalMapperSpec
            Return New TemporalMapperSpec With {
                .SpecVersion = CurrentSpecVersion,
                .MapperType = SpecType,
                .InputUnit = "unix_milliseconds",
                .InputRange = New Long() {InputMin, InputMax},
                .OutputRange = New Double() {OutputMin, OutputMax},
                .Clip = Clip
            }
        End Function

        Public Function ToJson() As String
            Return JsonSerializer.Serialize(Spec(), JsonOptions)
        End Function

        Public Shared Function FromJson(json As String) As TemporalRangeMapper
            Dim spec = JsonSerializer.Deserialize(Of TemporalMapperSpec)(json, JsonOptions)
            If spec Is Nothing Then
                Throw New InvalidOperationException("mapper spec JSON produced null.")
            End If

            Return FromSpec(spec)
        End Function

        Public Shared Function FromSpec(spec As TemporalMapperSpec) As TemporalRangeMapper
            If spec.SpecVersion <> CurrentSpecVersion Then
                Throw New ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", NameOf(spec))
            End If
            If spec.MapperType <> SpecType Then
                Throw New ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", NameOf(spec))
            End If
            If spec.InputUnit <> "unix_milliseconds" Then
                Throw New ArgumentException($"unsupported input_unit '{spec.InputUnit}'.", NameOf(spec))
            End If
            If spec.InputRange Is Nothing OrElse spec.InputRange.Length <> 2 Then
                Throw New ArgumentException("input_range must contain exactly two values.", NameOf(spec))
            End If
            If spec.OutputRange Is Nothing OrElse spec.OutputRange.Length <> 2 Then
                Throw New ArgumentException("output_range must contain exactly two values.", NameOf(spec))
            End If

            Return New TemporalRangeMapper(
                spec.InputRange(0),
                spec.InputRange(1),
                spec.OutputRange(0),
                spec.OutputRange(1),
                spec.Clip)
        End Function

        Private Function MapUnixMilliseconds(value As Long) As Double
            Dim v As Long = value
            If value < InputMin Then
                If Not Clip Then
                    Throw New ArgumentOutOfRangeException(NameOf(value), $"value {value} is below input_min {InputMin}; enable clip to clamp.")
                End If
                v = InputMin
            ElseIf value > InputMax Then
                If Not Clip Then
                    Throw New ArgumentOutOfRangeException(NameOf(value), $"value {value} is above input_max {InputMax}; enable clip to clamp.")
                End If
                v = InputMax
            End If

            Return OutputMin + ((CDbl(v - InputMin) / CDbl(InputMax - InputMin)) * (OutputMax - OutputMin))
        End Function
End Class

Public NotInheritable Class TemporalMapperSpec
        <JsonPropertyName("spec_version")>
        Public Property SpecVersion As String = CurrentSpecVersion

        <JsonPropertyName("mapper_type")>
        Public Property MapperType As String = "temporal_range"

        <JsonPropertyName("input_unit")>
        Public Property InputUnit As String = "unix_milliseconds"

        <JsonPropertyName("input_range")>
        Public Property InputRange As Long() = New Long(1) {}

        <JsonPropertyName("output_range")>
        Public Property OutputRange As Double() = New Double(1) {}

        <JsonPropertyName("clip")>
        Public Property Clip As Boolean
End Class

File: wrappers/csharp/LibRangeMap/TemporalRangeMapper.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LibRangeMap;

public sealed class TemporalRangeMapper
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        WriteIndented = false
    };

    private const string SpecVersion = MapperSpec.CurrentSpecVersion;
    private const string SpecType = "temporal_range";
    private const string InputUnit = "unix_milliseconds";

    public long InputMin { get; }
    public long InputMax { get; }
    public double OutputMin { get; }
    public double OutputMax { get; }
    public bool Clip { get; }
    public string? Name { get; }

    public TemporalRangeMapper(
        long inputMinUnixMilliseconds,
        long inputMaxUnixMilliseconds,
        double outputMin = -1.0,
        double outputMax = 1.0,
        bool clip = false,
        string? name = null)
    {
        ValidateFinite(nameof(outputMin), outputMin);
        ValidateFinite(nameof(outputMax), outputMax);

        if (outputMax <= outputMin)
        {
            throw new ArgumentException("output_range must be ordered from lower value to higher value.");
        }

        if (inputMaxUnixMilliseconds <= inputMinUnixMilliseconds)
        {
            throw new ArgumentException("input_range must be ordered from lower value to higher value.");
        }

        if (name is not null && name.Length == 0)
        {
            throw new ArgumentException("name must not be empty.", nameof(name));
        }

        InputMin = inputMinUnixMilliseconds;
        InputMax = inputMaxUnixMilliseconds;
        OutputMin = outputMin;
        OutputMax = outputMax;
        Clip = clip;
        Name = name;
    }

    public double MapValue(long unixMilliseconds) => MapUnixMilliseconds(unixMilliseconds);

    public double Map(long unixMilliseconds) => MapValue(unixMilliseconds);

    public double MapValue(DateTimeOffset value) => MapUnixMilliseconds(value.ToUnixTimeMilliseconds());

    public double Map(DateTimeOffset value) => MapValue(value);

    public double MapValue(DateTime value)
    {
        if (value.Kind == DateTimeKind.Unspecified)
        {
            throw new ArgumentException("DateTime with Kind=Unspecified is not supported; use UTC, Local, or DateTimeOffset.", nameof(value));
        }

        var utc = value.Kind == DateTimeKind.Utc ? value : value.ToUniversalTime();
        return MapValue(new DateTimeOffset(utc, TimeSpan.Zero));
    }

    public double Map(DateTime value) => MapValue(value);

    public TemporalMapperSpec Spec() => ToSpec();

    public string ToJson() => JsonSerializer.Serialize(ToSpec(), JsonOptions);

    public static TemporalRangeMapper FromJson(string json)
    {
        var spec = JsonSerializer.Deserialize<TemporalMapperSpec>(json, JsonOptions)
            ?? throw new InvalidOperationException("mapper spec JSON produced null.");
        return FromSpec(spec);
    }

    public static TemporalRangeMapper FromSpec(TemporalMapperSpec spec)
    {
        if (spec.SpecVersion != SpecVersion)
        {
            throw new ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", nameof(spec));
        }

        if (spec.MapperType != SpecType)
        {
            throw new ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", nameof(spec));
        }

        if (spec.InputUnit != InputUnit)
        {
            throw new ArgumentException($"unsupported input_unit '{spec.InputUnit}'.", nameof(spec));
        }

        if (spec.InputRange is null || spec.InputRange.Length != 2)
        {
            throw new ArgumentException("input_range must contain exactly two values.", nameof(spec));
        }

        if (spec.OutputRange is null || spec.OutputRange.Length != 2)
        {
            throw new ArgumentException("output_range must contain exactly two values.", nameof(spec));
        }

        return new TemporalRangeMapper(
            spec.InputRange[0],
            spec.InputRange[1],
            spec.OutputRange[0],
            spec.OutputRange[1],
            spec.Clip,
            spec.Name);
    }

    private double MapUnixMilliseconds(long value)
    {
        var v = (double)value;

        if (value < InputMin)
        {
            if (!Clip)
            {
                throw new ArgumentOutOfRangeException(nameof(value), $"value {v} is below input_range lower bound {InputMin}; enable clip to clamp.");
            }

            v = InputMin;
        }
        else if (value > InputMax)
        {
            if (!Clip)
            {
                throw new ArgumentOutOfRangeException(nameof(value), $"value {v} is above input_range upper bound {InputMax}; enable clip to clamp.");
            }

            v = InputMax;
        }

        return OutputMin + ((v - InputMin) / (InputMax - InputMin)) * (OutputMax - OutputMin);
    }

    private static void ValidateFinite(string label, double value)
    {
        if (double.IsNaN(value) || double.IsInfinity(value))
        {
            throw new ArgumentException($"{label} must be a finite number.");
        }
    }

    private TemporalMapperSpec ToSpec()
    {
        return new TemporalMapperSpec
        {
            SpecVersion = SpecVersion,
            MapperType = SpecType,
            InputUnit = InputUnit,
            InputRange = new[] { InputMin, InputMax },
            OutputRange = new[] { OutputMin, OutputMax },
            Clip = Clip,
            Name = Name
        };
    }
}

public sealed class TemporalMapperSpec
{
    [JsonPropertyName("spec_version")]
    public string SpecVersion { get; init; } = MapperSpec.CurrentSpecVersion;

    [JsonPropertyName("mapper_type")]
    public string MapperType { get; init; } = "temporal_range";

    [JsonPropertyName("input_unit")]
    public string InputUnit { get; init; } = "unix_milliseconds";

    [JsonPropertyName("input_range")]
    public long[] InputRange { get; init; } = new long[2];

    [JsonPropertyName("output_range")]
    public double[] OutputRange { get; init; } = new double[2];

    [JsonPropertyName("clip")]
    public bool Clip { get; init; }

    [JsonPropertyName("name")]
    public string? Name { get; init; }
}
